window/src/os/wayland/copy_and_paste.rs

```diff
@@ -35,14 +41,83 @@ impl CopyAndPaste {
         }
     }
 
-    pub fn get_clipboard_data(&mut self) -> anyhow::Result<FileDescriptor> {
-        let offer = self
-            .data_offer
-            .as_ref()
-            .ok_or_else(|| anyhow!("no data offer"))?;
-        let pipe = Pipe::new().map_err(Error::msg)?;
-        offer.receive(TEXT_MIME_TYPE.to_string(), pipe.write.as_raw_fd());
-        Ok(pipe.read)
+    pub fn get_clipboard_data(&mut self, clipboard: Clipboard) -> anyhow::Result<FileDescriptor> {
+        let conn = crate::Connection::get().unwrap().wayland();
+        let primary_selection = if let Clipboard::PrimarySelection = clipboard {
+            conn.pointer.primary_selection_device.as_ref()
+        } else {
+            None
+        };
+        match primary_selection {
+            Some(device) => {
+                let pipe = device.with_selection(|offer| {
+                    offer
+                        .ok_or_else(|| anyhow!("no primary selection offer"))
+                        .and_then(|o| {
+                            o.receive(TEXT_MIME_TYPE.to_string())
+                                .with_context(|| "failed to open read pipe".to_string())
+                        })
+                })?;
+                Ok(unsafe { FileDescriptor::from_raw_fd(pipe.into_raw_fd()) })
+            }
+            None => {
+                let offer = self
+                    .data_offer
+                    .as_ref()
+                    .ok_or_else(|| anyhow!("no data offer"))?;
+                let pipe = Pipe::new().map_err(Error::msg)?;
+                offer.receive(TEXT_MIME_TYPE.to_string(), pipe.write.as_raw_fd());
+                Ok(pipe.read)
+            }
+        }
+    }
+
+    pub fn set_clipboard_data(&mut self, clipboard: Clipboard, data: String) {
+        let conn = crate::Connection::get().unwrap().wayland();
+        let primary_selection = if let Clipboard::PrimarySelection = clipboard {
+            conn.environment
+                .borrow()
+                .get_primary_selection_manager()
+                .zip(conn.pointer.primary_selection_device.as_ref())
+        } else {
+            None
+        };
+
+        match primary_selection {
+            Some((manager, device)) => {
+                let source = PrimarySelectionSource::new(
+                    &manager,
+                    &[TEXT_MIME_TYPE.to_string()],
+                    move |event, _dispatch_data| match event {
+                        PrimarySelectionSourceEvent::Cancelled => {
+                            unimplemented!();
```

Comment:
this will crash the program if this ever triggers: can we log a warning and NOP here instead?

Reply:
Ah yeah, actually I wanted to implement this 🙈 Sorry totally forgot this when reshuffling all the code.

Thanks for catching this, I'll push an update to this PR.

Reply:
Added the missing piece :slightly_smiling_face: